dm/syncer/binlogstream/stream_modifier.go

```diff
@@ -0,0 +1,273 @@
+		zap.Stringer("new operator", toInject))
+	return nil
+}
+
+// Delete will delete an operator. `posStr` should be in the format of "binlog-file:pos".
+func (m *streamModifier) Delete(posStr string) error {
+	pos, err := binlog.PositionFromStr(posStr)
+	if err != nil {
+		return err
+	}
+
+	toDeleteIndex := 0
+	// not too much op, don't need binary search
+	for i, op := range m.ops {
+		if op.pos.Compare(pos) < 0 {
+			toDeleteIndex = i + 1
+		} else {
+			break
+		}
+	}
+
+	if toDeleteIndex < m.nextOp || toDeleteIndex == len(m.ops) {
+		return terror.ErrSyncerOperatorNotExist.Generate(posStr)
+	}
+	pre := m.ops[toDeleteIndex]
+	if pre.pos.Compare(pos) != 0 {
+		return terror.ErrSyncerOperatorNotExist.Generate(posStr)
+	}
+	copy(m.ops[toDeleteIndex:], m.ops[toDeleteIndex+1:])
+	m.ops = m.ops[:len(m.ops)-1]
+	return nil
+}
+
+// ListEqualAndAfter returns a JSON string of operators equals and after the given
+// position.
+// - if argument is "", it returns all operators.
+// - Otherwise caller should make sure the argument in format of "binlog-file:pos"
+//   and it returns all operators >= this position.
+func (m *streamModifier) ListEqualAndAfter(posStr string) string {
+	toPrint := []*operator{}
+	if posStr == "" {
+		toPrint = m.ops
+	} else {
+		pos, err := binlog.PositionFromStr(posStr)
+		if err != nil {
+			m.logger.DPanic("invalid position, should be verified in caller",
+				zap.String("position", posStr))
+			return ""
+		}
+
+		i := 0
+		for i < len(m.ops) && m.ops[i].pos.Compare(pos) < 0 {
+			i++
+		}
+
+		if i < len(m.ops) {
+			toPrint = m.ops[i:]
+		}
+	}
+
+	reqs := make([]*pb.HandleWorkerErrorRequest, 0, len(toPrint))
+	for _, op := range toPrint {
+		reqs = append(reqs, op.originReq)
+	}
+
+	b, err2 := json.Marshal(reqs)
+	if err2 != nil {
+		m.logger.DPanic("failed to marshal operators", zap.Error(err2))
+	}
+	return string(b)
+}
+
+// RemoveOutdated removes outdated operators which will not be triggered again after
+// upstream binlog streamer reset. A common usage is to use global checkpoint as
+// the argument.
+// RemoveOutdated will not remove the operator equals or after the `front`.
+func (m *streamModifier) RemoveOutdated(pos mysql.Position) {
+	lastRemoveIndex := 0
+	for i, op := range m.ops {
+		if op.pos.Compare(pos) < 0 {
+			lastRemoveIndex = i
+		} else {
+			break
+		}
+	}
+	if lastRemoveIndex > m.nextOp-1 {
+		lastRemoveIndex = m.nextOp - 1
+	}
+
+	m.ops = m.ops[lastRemoveIndex+1:]
+	m.nextOp -= lastRemoveIndex + 1
+}
+
+func (m *streamModifier) front() *operator {
+	if m.nextOp == len(m.ops) {
+		return nil
+	}
+	return m.ops[m.nextOp]
+}
+
+func (m *streamModifier) next() {
+	m.nextOp++
+	m.nextEventInOp = 0
+}
+
+// reset will also reset nextEventInOp to a correct value.
+func (m *streamModifier) reset(loc binlog.Location) {
+	m.nextOp = 0
+	m.nextEventInOp = 0
+
+	for i, op := range m.ops {
+		if op.pos.Compare(loc.Position) < 0 {
+			m.nextOp = i + 1
+		} else {
+			break
+		}
+	}
```

Comment:
quite common operation to find the first index >= some location, maybe abstract into one method

Reply:
https://github.com/pingcap/tiflow/pull/5890/commits/5f8f8a2b3fa2b06a3a0339b7ce8c0c54f5be1e39

Reply:
https://github.com/pingcap/tiflow/pull/5890/commits/00c29358a3420624f8150a2312da70603ab45b51